### ISSUE TITLE
Type parameters don't narrow correctly when used with `T.any`

### ISSUE THREAD

**Issue by gmalette** [sorbet/sorbet#2519]
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AGeneric%0A%0A%20%20Elem%20%3D%20type_member%0A%0A%20%20sig%20do%0A%20%20%20%20type_parameters(%3AT)%0A%20%20%20%20%20%20.params(%0A%20%20%20%20%20%20%20%20object_or_a%3A%20T.any(%0A%20%20%20%20%20%20%20%20%20%20A%5BT.type_parameter(%3AT)%5D%2C%0A%20%20%20%20%20%20%20%20%20%20T.type_parameter(%3AT)%0A%20%20%20%20%20%20%20%20)%0A%20%20%20%20%20%20)%0A%20%20%20%20%20%20.returns(A%5BT.type_parameter(%3AT)%5D)%0A%20%20end%0A%20%20def%20self.wrap(object_or_a)%0A%20%20%20%20T.reveal_type(T.type_parameter(%3AT))%0A%20%20%20%20case%20T.cast(object_or_a%2C%20BasicObject)%0A%20%20%20%20when%20A%0A%20%20%20%20%20%20T.cast(object_or_a%2C%20A%5BT.untyped%5D)%0A%20%20%20%20else%0A%20%20%20%20%20%20A%5BT.type_parameter(%3AT)%5D.new(object_or_a)%0A%20%20%20%20end%0A%20%20end%0A%0A%20%20sig%20%7B%20params(elem%3A%20Elem).void%20%7D%0A%20%20def%20initialize(elem)%0A%20%20end%0Aend%0A%0Adef%20main%0A%20%20a_integer%20%3D%20A.wrap(1)%0A%20%20a_from_wrapped%20%3D%20A.wrap(A%5BInteger%5D.new(1))%0A%20%20T.reveal_type(a_from_wrapped)%0Aend)

```ruby
# typed: true
class A
  extend T::Sig
  extend T::Generic

  Elem = type_member

  sig do
    type_parameters(:T)
      .params(
        object_or_a: T.any(
          A[T.type_parameter(:T)],
          T.type_parameter(:T)
        )
      )
      .returns(A[T.type_parameter(:T)])
  end
  def self.wrap(object_or_a)
    case T.cast(object_or_a, BasicObject)
    when A then T.cast(object_or_a, A[T.untyped])
    else
      A[T.type_parameter(:T)].new(object_or_a)
    end
  end

  sig { params(elem: Elem).void }
  def initialize(elem)
  end
end

def main
  a_integer = A.wrap(1)
  a_from_wrapped = A.wrap(A[Integer].new(1))
  T.reveal_type(a_from_wrapped)
end
```

#### Observed output

```
editor.rb:34: Revealed type: A[A[Integer]] https://srb.help/7014
    34 |  T.reveal_type(a_from_wrapped)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  From:
    editor.rb:33:
    33 |  a_from_wrapped = A.wrap(A[Integer].new(1))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^
Errors: 1
```

#### Expected behavior

I expect the revealed type to be `A[Integer]`

In the code example, the function can be expressed as 

```
fn wrap<T>(x: T | A<T>) -> A<T>
```

However, Sorbet doesn't seem to unwrap the `A[T]` from the union type, and returns an `A[A[T]]`

**Comment by DarkDimius** [comment]
The solution that typechecker chose is with T = `A[Integer]` which does correctly solve the signature.
Note that `T.any` is unordeded and there's no guarantee on what side of it does the solver decide to solve for.

**Comment by gmalette** [comment]
@DarkDimius should the solver not pick the more precise type? If it doesn't, AFAIK there is no way to express that method. It's not possible to express it at the caller either, because Sorbet doesn't allow parameterizing the type like `A[Integer].wrap(1)` for methods other than `new` (correct me if I'm wrong).

**Comment by DarkDimius** [comment]
> should the solver not pick the more precise type? 

Neither of the solution types is more precise than the other(`A[Integer]` and `Integer` are unrelated).

> because Sorbet doesn't allow parameterizing the type like A[Integer].wrap(1) for methods other than new (correct me if I'm wrong).

https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).returns(T.attached_class)%7D%0A%20%20def%20self.bar(x)%0A%20%20%20%20self.new()%0A%20%20end%0Aend%0A%0Adef%20main%0A%20T.reveal_type(A.bar(1))%0Aend

**Comment by gmalette** [comment]
`A[T]` is more precise than `T` tho

> https://sorbet.run/#%23%20typed%3A%20true%0Aclass%20A%0A%20%20extend%20T%3A%3ASig%0A%0A%20%20sig%20%7Bparams(x%3A%20Integer).returns(T.attached_class)%7D%0A%20%20def%20self.bar(x)%0A%20%20%20%20self.new()%0A%20%20end%0Aend%0A%0Adef%20main%0A%20T.reveal_type(A.bar(1))%0Aend

I don't understand how to take that example and convert it in the type parameter that accepts an `A[Integer]` or an `Integer` and returns an `A[Integer]`, can you explain what you meant please?

**Comment by DarkDimius** [comment]
> A[T] is more precise than T tho

What's your definition of precise? The one that I use is subtyping. From subtyping perspective these two types are unrelated

**Comment by DarkDimius** [comment]
> convert it in the type parameter that accepts an A[Integer] or an Integer and returns an A[Integer], 

You'll need to constrain T somehow so that identity-like signature isn't a valid solution.

Why do you want this to be a single method? From both code readability and performance standpoint in Stripe codebase it would be preferable to split this method on two

**Comment by gmalette** [comment]
> What's your definition of precise? 

Semantics, first, but also category:

"Anything"
vs
"An `A` parameterized by anything"

The 2nd one is a subset of the 1st one, less values will belong in it, and all values accepted in the 2nd one are accepted in the 1st one.

> Why do you want this to be a single method?

Short answer, Ruby doesn't have method overloading. In languages with overload, I would've written this `Constraint<Integer>.wrap(value)`.

We're adding typing to our constraint library. The constraints are very abstract from the users of the code. Without going into the details, imagine they can have either `Integer(1)` or `LessThan(Integer(2))`. `Constraint.wrap` will return `Exactly(Integer(1))` and `LessThan(Integer(2))` respectively, such that the solver can do its job correctly. Now you're probably asking why we can't just use `Integer` directly, and it's because we need to use different values for the types handled natively by the library and the type defined by the users. 

In short, it's a matter of ergonomics, similar to [`Array()`](https://ruby-doc.org/core-2.7.0/Kernel.html#method-i-Array) method.
